### ISSUE TITLE
Assigning callable to attribute is generating typing error when attribute is called.

### ISSUE THREAD

**Issue by philosofool** [microsoft/pyright#7303]
**Describe the bug**
Pyright gives error:
```
Argument of type "Literal[2]" cannot be assigned to parameter of type "A@__init__"
  Type "Literal[2]" cannot be assigned to type "A@__init__"Pylance[reportArgumentType](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportArgumentType)
```
when using the following code:
```python

A = TypeVar('A')

class LambdaReadableDataSource:
    """Provide data source with read_key method provided by pure function."""
    def __init__(self, read_key: Callable[[str], A]):
        self.read_key = read_key

class LambdaDatasource(LambdaReadableDataSource):
    def __init__(self, read_key: Callable[[str], A], write_key: Callable[[str, A], None]):
        super().__init__(read_key)
        self.write_key = write_key

def TestLambdaDataSource():
    data = {}

    def write(k: str, v: int) -> None:
        data[k] = v

    def read(k: str) -> int:
        return data[k]

    lambda_source = LambdaDataSource(read, write)
    assert lambda_source.write_key('2', 2) is None  # this line raises the error. The integer `2` has the error "squiggle".
    assert lambda_source.read_key('2') == 2

```
The expected behavior would be to raise a typing error with the type of `v` in write_key did not match the type of the signature of  in `write`, but not otherwise.

**VS Code extension or command-line**
Running as VSCode extension, through Pylance installation, v.2024.2.2

**Comment by erictraut** [comment]
You are using type variables incorrectly here. I suspect that your intent is for `LambdaReadableDataSource` and `LambdaDataSource` to be generic classes whose instance variables `read_key` and `write_key` have a generic type. Instead, you've defined non-generic classes with generic `__init__` methods. In other words, the type variable `A` is not scoped to the class but instead to the `__init__` method only. Since you haven't provided a type declaration for instance variables `read_key` and `write_key`, their types are inferred from the assignment statements, but the assigned values have a type that is a method-scoped generic.

If you change your code as follows, it will work the way you expect:

```python
class LambdaReadableDataSource(Generic[A]):
    def __init__(self, read_key: Callable[[str], A]):
        self.read_key = read_key

class LambdaDataSource(LambdaReadableDataSource[A]):
    def __init__(
        self, read_key: Callable[[str], A], write_key: Callable[[str, A], None]
    ):
        super().__init__(read_key)
        self.write_key = write_key
```

**Comment by philosofool** [comment]
Thank you. Your suggestions do fix the issue.
